### PR TITLE
ModuleDuplicator & ComponentPrinter now benefit from upgraded parts

### PR DESCRIPTION
**Pull request by SyncIt21**
**About the pull request**

Module Duplicator & Component Printer now provide more local storage and cheaper production costs when upgraded with better matter bins & manipulators respectively. Previously this wasn't the case

**Why its good for the game**
 Better parts mean lower costs and more storage. Thats always a good thing

**Change Log**
🆑 
add: Module Duplicator & Component Printer now get extra storage & reduced costs from more efficient parts
/:cl:

### CODE REVIEW

**Comment by Sealed101** [comment]
tying the normal servers to the master one will make RnD completely halt when an antagonist takes out the master server, since the station cannot rebuild it without admin intervention.
wasnt there a veto or something placed on upgrades improving rnd servers some time back

**Comment by SyncIt21** [comment]
The master server must have some importance in making all other server's work. If life continues as normal when the master is destroyed that kinda makes the master meaningless in its name.
 We could have a spare master server circuit board inside the R & D director office to rebuild it if it does get destroyed or just as 5ge admin to make a new console it isn't a big deal
Improving R & D points should be allowed as it logically makes sense with better parts = more points we can decide what that increment should be but I disagree with that veto of no upgrades at all

**Comment by RikuTheKiller** [comment]
> The master server must have some importance in making all other server's work. If life continues as normal when the master is destroyed that kinda makes the master meaningless in its name. We could have a spare master server circuit board inside the R & D director office to rebuild it if it does get destroyed or just ask admin to make a new console it isn't a big deal Improving R & D points should be allowed as it logically makes sense with better parts = more points we can decide what that increment should be but I disagree with that veto of no upgrades at all

1. Players should NEVER have to contact admins for normal behaviour in a match. That's just dumb. It also breaks the 4th wall which wouldn't be very suitable for MRP. (Oh wow the server appeared out of thin air!)

2. They could simply get rid of the spare and also there are lots of lowpop rounds with no RD at all so having to hack shit just to repair the master server after FriendlyAntag201 destroys it is annoying as fuck.

3. Just make the master server's contribution to point generation like, three times that of other servers. We could simply assume that it's a more advanced server meant for more difficult data crunching tasks. So the crew can replace it by building 3 servers in it's place. (I don't like permanent, hardcoded changes. Almost everything should be able to be restored in some way, at least in my opinion.)

4. Or just add some way of rebuilding the master server. Perhaps by inserting a copy of the master hard drive while making it. Get a copy from the RD's console or an R&D console. Not sure which is more suitable for this. (I'd pick R&D console, but you do you.)

**Comment by Mothblocks** [comment]
your changelog is pretty useless, and also tweak does not exist anymore. detail the changes in a way a player will understand exactly what changed

**Comment by SyncIt21** [comment]
> > The master server must have some importance in making all other server's work. If life continues as normal when the master is destroyed that kinda makes the master meaningless in its name. We could have a spare master server circuit board inside the R & D director office to rebuild it if it does get destroyed or just ask admin to make a new console it isn't a big deal Improving R & D points should be allowed as it logically makes sense with better parts = more points we can decide what that increment should be but I disagree with that veto of no upgrades at all
> 
> 1. Players should NEVER have to contact admins for normal behaviour in a match. That's just dumb. Not to mention that admins aren't on all of the time, especially on Basil and Sybil. It also breaks the 4th wall which wouldn't be very suitable for MRP. (Oh wow the server appeared out of thin air!)
> 2. They could simply get rid of the spare and also there are lots of lowpop rounds with no RD at all so having to hack shit just to replace the master server after FriendlyAntag201 destroys it is annoying as fuck.
> 3. Just make the master server's contribution to point generation like, three times that of other servers. We could simply assume that it's a more advanced server meant for more intensive data crunching tasks. This means the crew can replace it by building another 3 servers in it's place. (I don't like permanent, hardcoded changes. Almost everything should be able to be restored in some way, at least in my opinion.)
> 4. Or just add some way of rebuilding the master server. Perhaps by inserting a copy of the master hard drive while making it. Get a copy from the RD's console or an R&D console. Not sure which is more suitable for this. (I'd pick R&D console, but you do you.)

@RikuTheKiller Those are some good points. I made some modifications now

**Comment by private-tristan** [comment]
The changelog on this seems a bit cursed due to the videos and changelog itself mentioning videos (does the ingame changelog even support embeds?)

**Comment by SyncIt21** [comment]
> The changelog on this seems a bit cursed due to the videos and changelog itself mentioning videos (does the ingame changelog even support embeds?)

I have moved videos up away from the changelog. Hopefully i should see the correct labels on this PR now

**Comment by moocowswag** [comment]
Extremely based pr, totally healthy for the game.

**Comment by RikuTheKiller** [comment]
> Extremely based pr, totally healthy for the game.
> 
> Finally RDs have an actual job to do other than fill in unused sci jobs or press buttons on robot bondage console.

And maybe getting basic medical research won't take 40 minutes since science has decided to focus on their mechs.

**Comment by moocowswag** [comment]
> And maybe getting basic medical research won't take 40 minutes since science has decided to focus on their mechs.
> 
> It'll still happen, but the issue can be fixed by upgrading the servers, at least.

shhhhhhhh dont say anything about getting research faster.

**Comment by RikuTheKiller** [comment]
> > And maybe getting basic medical research won't take 40 minutes since science has decided to focus on their mechs.
> > It'll still happen, but the issue can be fixed by upgrading the servers, at least.
> 
> shhhhhhhh dont say anything about getting research faster, orange himself said the intended round duration is 90 minutes, we should only finish the tech tree long after 40 minutes.

You mean getting basic medical research after 40 minutes, not finishing the tech tree. :)

File: code/controllers/subsystem/research.dm [hunk 1]
```diff
@@ -86,17 +86,46 @@ SUBSYSTEM_DEF(research)
...
 /datum/controller/subsystem/research/fire()
+	var/can_synchronize=FALSE
+
+	//since the game will ever have only 1 master server we just return if that is offline
+	for(var/obj/machinery/rnd/server/master/master_server in master_servers)
+		can_synchronize=TRUE
+		if(!master_server.working)
+			///we still want to initialize the last time this method was called to prevent point gaines during the time the server was off
+			last_income = world.time
+			return
```
Comment (Watermelon914) [inline code/controllers/subsystem/research.dm:97]: this isn't a bugfix, it's a balance change.
R&D is supposed to work as long as there's one server available. The master server is there for traitor objectives

Comment (SyncIt21) [reply code/controllers/subsystem/research.dm:97]: Your right the bug fix isn't about the server its about component printer and it isn't related .

 I just added master server reliance as a feature and as I said in my PR servers depend on master on/off state only if the master server exists.

If master is destroyed then we go back to the old behavior of requiring only one server for functioning and points still continue to be generated at half the rate. plzz look at video 3

**Comment by SyncIt21** [comment]
Any maintainer wants to discuss about this PR getting merged? I have no pending changes here

File: code/controllers/subsystem/research.dm (hunk 1 above)
Comment (Watermelon914) [reply code/controllers/subsystem/research.dm:97]: how so? If the master server isn't working, it early returns.
If there is no master server, it works? This is inconsistent

**Review by Watermelon914 — CHANGES_REQUESTED** [review]
I don't really support the master server being disabled disabling other servers, also the boosted research points is not really that great either.
Can you just leave the changes with the ModuleDuplicator and ComponentPrinter in and remove all other changes from this PR?

File: code/controllers/subsystem/research.dm (hunk 1 above)
Comment (SyncIt21) [reply code/controllers/subsystem/research.dm:97]: Alright I'll remove the dependence

**Comment by SyncIt21** [comment]
> I don't really support the master server being disabled disabling other servers, also the boosted research points is not really that great either. Can you just leave the changes with the ModuleDuplicator and ComponentPrinter in and remove all other changes from this PR?

In the video I only showed a maximum 140+ point gain cause I had only tier 1 parts. but in reality if you have 3 servers with tier 4 parts you can get close to 300+ points. 
It's a big deal alright

**Comment by SyncIt21** [comment]
> I don't really support the master server being disabled disabling other servers, also the boosted research points is not really that great either. Can you just leave the changes with the ModuleDuplicator and ComponentPrinter in and remove all other changes from this PR?

I'll remove the dependence thats fine.  However the master should still be online to support synchronization which as I said above is a big deal

**Comment by RikuTheKiller** [comment]
> > I don't really support the master server being disabled disabling other servers, also the boosted research points is not really that great either. Can you just leave the changes with the ModuleDuplicator and ComponentPrinter in and remove all other changes from this PR?
> 
> I'll remove the dependence that's fine however point gains are big as I'll explain below
> 
> In the video I only showed a maximum 140+ point gain cause I had only tier 1 parts. but in reality if you have 3 servers with tier 4 parts you can get close to 400+ points.
> 
> From then onwards it gets more difficult, you have to create 7 server's for 600+ points, 12 server's for 900+ points etc. Check the formula with each server added its contribution is divided by N[N is the index of the server in the list so for the 5th server it would be divided by 5 and so on]
> 
> Its not a power creep either cause even with tier 1 scanning modules you have to cool the tile on which the server lies to around 70-80k and with tier 4 to 20-30k with freezers and they cant just build it in the open so if they hope to build this many server's they need engineering to setup an dedicated room for it,thus giving them some work to do.
> 
> Either that or they can just build 3 server's in the existing server room and gain 400+ points. A single tier 4 server is equivalent to the point gain in my video I.e 140+ points
> 
> It's a big deal alright. I hope you will consider this

I still support the whole upgrading thing. I've never liked the static research point system.

**Comment by Watermelon914** [comment]
I'm fine with the changes to the module printer and duplicate printer, but I don't support the changes to research so I'll probably not be merging this PR. If you can find a maintainer who has an interest in this idea, then good enough, but if you want the module printer and duplicate printer changes merged, you would be better off atomizing the PR so I can merge it separately.

I'm not saying that I'm going to block your idea, just that I'm indifferent to it and leaning onto the side of not really for it, considering it's a major balance change to research.

**Comment by SyncIt21** [comment]
It is a major balance change in all the good ways

> I'm fine with the changes to the module printer and duplicate printer, but I don't support the changes to research so I'll probably not be merging this PR. If you can find a maintainer who has an interest in this idea, then good enough, but if you want the module printer and duplicate printer changes merged, you would be better off atomizing the PR so I can merge it separately.
> 
> I'm not saying that I'm going to block your idea, just that I'm indifferent to it and leaning onto the side of not really for it, considering it's a major balance change to research.

it is a major balance change in all the good ways. Everyone in the chat supports it and you aren't giving good on why are you are indifferent to it. 

Splitting the PR at this point is too much work considering how old this is. It's much easier to merge all changes at once here.

**Comment by SyncIt21** [comment]
@Time-Green I would greatly appreciate your input & thoughts on this.

**Comment by Time-Green** [comment]
@SyncIt21 Sorry, I'm not merging this either, I have no balance feel for RnD. Just excise the balance change and you can get it merged. Putting the balance change in another PR (if you want that added) also means that it's a lot easier for people to find and respond to, instead of somewhat being hidden away

**Comment by SyncIt21** [comment]
> @SyncIt21 Sorry, I'm not merging this either, I have no balance feel for RnD. Just excise the balance change and you can get it merged. Putting the balance change in another PR (if you want that added) also means that it's a lot easier for people to find and respond to, instead of somewhat being hidden away

😭 

I guess i have no choice then.  Sure i will move the balance thing into its own PR which means only upgrading component printer & module printer will exist in this PR.

**Comment by SyncIt21** [comment]
@Watermelon914 , @Time-Green the point upgrades are gone :( . Can anyone review & merge this now?

File: code/modules/wiremod/core/component_printer.dm [hunk 2]
```diff
@@ -470,11 +522,11 @@
 		return TRUE
 	return default_deconstruction_screwdriver(user, "module-fab-o", "module-fab-idle", tool)
 
-/obj/machinery/module_duplicator/proc/get_material_cost_data(list/materials)
+/obj/machinery/module_duplicator/proc/get_material_cost_data(list/materials,efficiency_coeff)
```
Comment (Time-Green) [inline code/modules/wiremod/core/component_printer.dm:525]: Please use spacing

File: code/modules/wiremod/core/component_printer.dm [hunk 3]
```diff
@@ -445,11 +495,13 @@
...
 	var/index = 1
 
+	var/efficiency_coeff=calculate_efficiency()
+
 	for (var/list/design as anything in scanned_designs)
 		designs["[index]"] = list(
 			"name" = design["name"],
 			"desc" = design["desc"],
-			"cost" = get_material_cost_data(design["materials"]),
+			"cost" = get_material_cost_data(design["materials"],efficiency_coeff),
```
Comment (Time-Green) [inline code/modules/wiremod/core/component_printer.dm:504]: Spacing

File: code/modules/wiremod/core/component_printer.dm [hunk 4]
```diff
@@ -337,12 +385,14 @@
 				say("Mineral access is on hold, please contact the quartermaster.")
 				return TRUE
 
-			if (!materials.mat_container?.has_materials(design["materials"]))
+			var/efficiency_coeff=calculate_efficiency()
```
Comment (Time-Green) [inline code/modules/wiremod/core/component_printer.dm:388]: Here too

File: code/modules/wiremod/core/component_printer.dm [hunk 5]
```diff
@@ -184,11 +210,11 @@
 		return TRUE
 	return default_deconstruction_screwdriver(user, "fab-o", "fab-idle", tool)
 
-/obj/machinery/component_printer/proc/get_material_cost_data(list/materials)
+/obj/machinery/component_printer/proc/get_material_cost_data(list/materials,efficiency_coeff)
 	var/list/data = list()
 
 	for (var/datum/material/material_type as anything in materials)
-		data[initial(material_type.name)] = materials[material_type]
+		data[initial(material_type.name)] = materials[material_type]*efficiency_coeff
```
Comment (Time-Green) [inline code/modules/wiremod/core/component_printer.dm:217]: Ditto

**Review by Time-Green — CHANGES_REQUESTED** [review]
Looks good! Please use spacing when typing paramters though and assigning stuff to variables
So 
A=B to A = B
or
/proc/do_stuff(atom/A,atom/B) to /proc/do_stuff(atom/A, atom/B)

I've marked a few instances in your code as example

File: code/modules/wiremod/core/component_printer.dm (hunk 2 above)
Comment (SyncIt21) [reply code/modules/wiremod/core/component_printer.dm:525]: I'm on it

File: code/modules/wiremod/core/component_printer.dm (hunk 3 above)
Comment (SyncIt21) [reply code/modules/wiremod/core/component_printer.dm:504]: Im getting good at this :(

File: code/modules/wiremod/core/component_printer.dm (hunk 4 above)
Comment (SyncIt21) [reply code/modules/wiremod/core/component_printer.dm:388]: EEEEEEEE

File: code/modules/wiremod/core/component_printer.dm (hunk 5 above)
Comment (SyncIt21) [reply code/modules/wiremod/core/component_printer.dm:217]: OOOOOO

File: code/modules/wiremod/core/component_printer.dm (hunk 3 above)
Comment (Time-Green) [reply code/modules/wiremod/core/component_printer.dm:504]: Don't feel bad :(

**Comment by SyncIt21** [comment]
@Time-Green Done :)

**Pull request merged**